File: src/reverse_linked_list_206/solution_recursive.go
package reverse_linked_list_206

func reverseListRecursive(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	var newHead = reverseListRecursive(head.Next)
	head.Next.Next = head
	head.Next = nil

	return newHead
}

func TestRecursive() {
	head := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
				Next: &ListNode{
					Val: 4,
					Next: &ListNode{
						Val:  5,
						Next: nil,
					},
				},
			},
		},
	}

	newHead := reverseListRecursive(head)
	printList(newHead)
}

File: src/reverse_linked_list_206/solution_recursive.ts
// Definition for singly-linked list.

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function reverseList(head: ListNode | null): ListNode | null {
  if (head == null || head.next == null) {
    return head;
  }

  const newHead = reverseList(head.next);
  head.next.next = head;
  head.next = null;
  return newHead;
}

function printList(head: ListNode | null) {
  let current: ListNode | null = head;

  while (current !== null) {
    process.stdout.write(`${current.val} -> `);
    current = current.next;
  }
}

export function test() {
  const head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

  const newHead = reverseList(head);
  printList(newHead);
}
